spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,360 @@
+    counterpartyUpgradeStatus: UpgradeStatus,
+    proofConnection: CommitmentProof,
+    proofUpgradeStatus: CommitmentProof,
+    proofHeight: Height
+) {
+    // current connection must be OPEN or UPGRADE_INIT (crossing hellos)
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(connection.state == (OPEN || UPGRADE_INIT))
+
+    // abort transaction if an unmodifiable field is modified
+    // upgraded connection state must be in `UPGRADE_TRY`
+    abortTransactionUnless(
+        proposedUpgrade.connection.state == UPGRADE_TRY &&
+        proposedUpgrade.connection.counterpartyConnectionIdentifier == currentConnection.counterpartyConnectionIdentifier &&
+        proposedUpgrade.connection.clientIdentifier == currentConnection.clientIdentifier &&
+        proposedUpgrade.connection.counterpartyClientIdentifier == currentConnection.counterpartyClientIdentifier &&
+    )
+
+    // either timeout height or timestamp must be non-zero
+    abortTransactionUnless(proposedUpgrade.TimeoutHeight != 0 || proposedUpgrade.TimeoutTimestamp != 0)
+
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(currentConnection.client.verifyConnectionState(proofHeight, proofConnection, counterpartyConnection))
+    abortTransactionUnless(currentConnection.client.verifyUpgradeStatus(proofHeight, proofUpgradeStatus, counterpartyUpgradeStatus))
+
+    // verify that counterparty connection unmodifiable fields have not changed and counterparty state
+    // is UPGRADE_INIT
+    restoreConnectionUnless(
+        counterpartyConnection.state == UPGRADE_INIT &&
+        counterpartyConnection.counterpartyConnectionIdentifier == identifier &&
+        counterpartyConnection.clientIdentifier == currentConnection.counterpartyClientIdentifier &&
+        counterpartyConnection.counterpartyClientIdentifier == currentConnection.clientIdentifier
+    )
+    restoreConnectionUnless(counterpartyUpgradeStatus.state == UPGRADE_INIT)
+
+    // counterparty-specified timeout must not have exceeded
+    restoreConnectionUnless(
+        counterparty.UpgradeStatus.TimeoutHeight < currentHeight() &&
+        counterparty.UpgradeStatus.TimeoutTimestamp < currentTimestamp()
+    )
+
+    // verify chosen versions are compatible
+    versionsIntersection = intersection(counterpartyConnection.version, proposedUpgrade.Connection.version)
+    version = pickVersion(versionsIntersection) // throws if there is no intersection
+
+    // both connection ends must be mutually compatible.
+    restoreConnectionUnless(IsCompatible(counterpartyConnection, proposedUpgrade.Connection))
+
+    upgradeStatus = UpgradeStatus{
+        state: UPGRADE_TRY,
+        timeoutHeight: proposedUpgrade.timeoutHeight,
+        timeoutTimestamp: proposedUpgrade.timeoutTimestamp,
+    }
+
+    provableStore.set(statusPath(identifier), upgradeStatus)
+    provableStore.set(connectionPath(identifier), proposedUpgrade.connection)
```

Comment:
Interesting, I'd think you'd want to leave the connection unchanged until the upgrade process successfully completes. Otherwise you could potentially disrupt packet flow if the connection is in UPGRADE_INIT instead of OPEN?

I'd expect you to store the proposed upgrade connection under a proposed upgrade connection path. Use that for proofs and then if successful update the connection on ack/confirm?

Reply:
I don't want the case where chainA has its connectionEnd OPEN with the old parameters, and chainB has its connectionEnd OPEN with the new parameters as that may cause unexpected consequences.

If a connection is `OPEN` on both ends it should be a valid connection always.

So either both ends are `OPEN` running the old parameters, or both ends are `OPEN` running the new parameters.

The only way to accomplish this is to temporarily close the connection on both sides (INIT and TRY), then reopen after the upgrade is complete on either end (ACK and CONFIRM).

This does mean packet flow is temporarily halted but i believe this is fine. Once the connection is back up, the relayers can either receive or timeout the pending packets

Reply:
This is also why I specify a timeout, because I don't want the "temporary pause" to last arbitrarily long

Reply:
yeah, this makes sense to me